### PR TITLE
consensus/pbft: design pbft config

### PR DESCRIPTION
**Pull request by bailantaotao**

### CODE REVIEW

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -410,11 +410,31 @@ var (
...
 	// PBFT settings
-	PBFTTimeoutFlag = cli.IntFlag{
+	PBFTTimeoutFlag = cli.Uint64Flag{
 		Name:  "pbft_timeout",
 		Usage: "The timeout for each PBFT round in millisecond",
 		Value: 2000,
 	}
+	PBFTBlockPeriod = cli.Uint64Flag{
+		Name:  "pbft_block_period",
+		Usage: "Default minimum difference between two consecutive block's timestamps. (milliseconds)",
```
Comment (markya0616) [inline cmd/utils/flags.go]: block time is second-based. suggest to use second.

File: cmd/utils/flags.go [hunk 2]
```diff
@@ -410,11 +410,31 @@ var (
...
+		Value: 1000,
+	}
+	PBFTBlockPauseTime = cli.Uint64Flag{
+		Name:  "pbft_block_pause_time",
+		Usage: "Pause time when zero tx in previous block. (milliseconds)",
```
Comment (markya0616) [inline cmd/utils/flags.go]: use second?

File: cmd/utils/flags.go [hunk 3]
```diff
@@ -410,11 +410,31 @@ var (
...
+		Value: 100,
```
Comment (markya0616) [inline cmd/utils/flags.go]: this default value should be larger than `PBFTBlockPeriod`

File: consensus/pbft/config.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package pbft
+
+type ProposerPolicy int
+
+const (
+	PBFT ProposerPolicy = iota
+	RoundRobin
```
Comment (markya0616) [inline consensus/pbft/config.go]: what's the differences between PBFT and RoundRobin?

File: eth/backend.go [hunk 5]
```diff
@@ -214,7 +215,15 @@ func CreateConsensusEngine(ctx *node.ServiceContext, config *Config, chainConfig
...
 	if chainConfig.PBFT != nil {
-		return pbftBackend.New(config.PBFTTimeout, ctx.EventMux, ctx.NodeKey(), db)
+		pbftConfig := &pbft.Config{
+			RequestTimeoutMsec: config.PBFTTimeout,
+			BlockPeriodMsec:    config.PBFTBlockPeriod,
+			BlockPauseTimeMsec: config.PBFTBlockPauseTime,
+			ProposerPolicy:     pbft.ProposerPolicy(config.PBFTProposerPolicy),
+			CheckPointPeriod:   config.PBFTCheckPointPeriod,
+		}
+		log.Info("config", "gg", pbftConfig)
```
Comment (markya0616) [inline eth/backend.go]: gg XD

File: cmd/utils/flags.go [hunk 6]
```diff
@@ -410,11 +410,31 @@ var (
...
+	}
+	PBFTProposerPolicy = cli.Uint64Flag{
+		Name:  "pbft_proposer_policy",
+		Usage: "The policy for proposer, the detail is not determined",
+		Value: 0,
```
Comment (markya0616) [inline cmd/utils/flags.go]: can we set the value by `DefaultConfig`?

File: consensus/pbft/config.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+	BlockPeriod        uint64         // Default minimum difference between two consecutive block's timestamps in second
+	BlockPauseTime     uint64         // Pause time when zero tx in previous block, , values should be larger than pbft_block_period
+	ProposerPolicy     ProposerPolicy // The policy for proposer, the detail is not determined
+	CheckPointPeriod   int            // Synchronizes the mapping's checkpoint to the blocks on each round
+}
+
+var DefaultConfig = &Config{
+	RequestTimeoutMsec: 2000,
+	BlockPeriod:        1,
+	BlockPauseTime:     1,
```
Comment (bailantaotao) [inline consensus/pbft/config.go]: lager than BlockPeriod

**Comment by bailantaotao** [comment]
all comments are fixed in new commits

File: consensus/pbft/core/core.go [hunk 8]
```diff
@@ -61,34 +61,36 @@ type Engine interface {
 	Stop() error
 }
 
-func New(backend pbft.Backend) Engine {
+func New(backend pbft.Backend, requestTimeout uint64) Engine {
```
Comment (alanchchen) [inline consensus/pbft/core/core.go]: How about passing `*pbft.Config` here?
We might have more configurations.

Comment (bailantaotao) [inline consensus/pbft/core/core.go]: OK

Comment (bailantaotao) [inline consensus/pbft/core/core.go]: fixed in c2ab80e

File: consensus/pbft/config.go [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+type ProposerPolicy int
+
+const (
+	RoundRobin ProposerPolicy = iota
+)
+
+type Config struct {
+	RequestTimeoutMsec uint64         // The timeout for each PBFT round
+	BlockPeriod        uint64         // Default minimum difference between two consecutive block's timestamps in second
+	BlockPauseTime     uint64         // Pause time when zero tx in previous block, , values should be larger than pbft_block_period
```
Comment (tailingchen) [inline consensus/pbft/config.go]: remove redundant comma
`Pause time when zero tx in previous block, , values ...`

**Pull request merged**